File: elixir/day4/day4.ex
defmodule Day4 do
  def new_card(card, currNum, parseW) do
    if parseW === true do
      put_in(card, [:winning, currNum], currNum)
    else
      put_in(card, [:nums, currNum], currNum)
    end
  end

  def parse_row(row) do
    # row, currNum, card 
    # %{ cardId: int, nums: %{int,int}, winning: %{int, int}}
    parse_row(row, nil, %{nums: %{}, winning: %{}}, false)
  end

  def parse_row(<<num, rest::binary>>, nil, card, parseW) when num in ?0..?9 do
    parse_row(rest, num - ?0, card, parseW)
  end

  def parse_row(<<num, rest::binary>>, currNum, card, parseW) when num in ?0..?9 do
    parse_row(rest, currNum * 10 + num - ?0, card, parseW)
  end

  def parse_row(<<meh, rest::binary>>, nil, card, parseW) do
    if meh == 124 do
      parse_row(rest, nil, card, true)
    else
      parse_row(rest, nil, card, parseW)
    end
  end

  def parse_row(<<meh, rest::binary>>, currNum, card, parseW) when meh not in ?0..?9 do
    newCard = new_card(card, currNum, parseW)

    if meh == 124 do
      parse_row(rest, nil, newCard, true)
    else
      parse_row(rest, nil, newCard, parseW)
    end
  end

  def parse_row(<<"">>, currNum, card, p) do
    new_card(card, currNum, p)
  end

  def parse_row(nil, nil, card, _p) do
    card
  end

  def parse_id(line) do
    parse_id(line, nil)
  end

  def parse_id(<<char, rest::binary>>, nil) when char not in ?0..?9 do
    parse_id(rest, nil)
  end

  def parse_id(<<int, rest::binary>>, nil) when int in ?0..?9 do
    parse_id(rest, int - ?0)
  end

  def parse_id(<<int, rest::binary>>, id) when int in ?0..?9 do
    parse_id(rest, id * 10 + int - ?0)
  end

  def parse_id(<<>>, id) do
    id
  end

  def calculate_winning(card) do
    wins = card.winning |> Map.keys()
    calculate_winning(wins, card.nums, 0, 0, 0)
  end

  def calculate_winning([key | rest], nums, points, total, num_of_matches) do
    if Map.has_key?(nums, key) and key !== nil do
      if total == 0 do
        calculate_winning(rest, nums, points + 1, total + 1, num_of_matches + 1)
      else
        calculate_winning(rest, nums, points * 2, total + 1, num_of_matches + 1)
      end
    else
      calculate_winning(rest, nums, points, total, num_of_matches)
    end
  end

  def calculate_winning([], _nums, points, _total, num_of_matches) do
    {points, num_of_matches}
  end

  def update_my_copies(copies, num) do
    if Map.has_key?(copies, num) do
      Map.update!(copies, num, fn x -> x + 1 end)
    else
      Map.put(copies, num, 1)
    end
  end

  def add_winning(copies, copies_of_me, idx, winning_num, my_id) do
    if idx > winning_num do
      copies
    else
      curr_id = idx + my_id

      if Map.has_key?(copies, curr_id) do
        add_winning(
          Map.update!(copies, curr_id, fn x -> x + copies_of_me end),
          copies_of_me,
          idx + 1,
          winning_num,
          my_id
        )
      else
        add_winning(
          Map.put(copies, curr_id, copies_of_me),
          copies_of_me,
          idx + 1,
          winning_num,
          my_id
        )
      end
    end
  end

  def cal_copies(copies, winning_cards, idx) when idx > length(winning_cards) + 1 do
    copies
  end

  def cal_copies(copies, winning_cards, idx) do
    new_curr = Enum.at(winning_cards, idx)

    if new_curr == nil do
      copies
    else
      my_copies = update_my_copies(copies, new_curr.id)

      if new_curr.num_of_matches == 0 do
        cal_copies(my_copies, winning_cards, idx + 1)
      else
        new_copies =
          add_winning(
            my_copies,
            Map.get(my_copies, new_curr.id),
            1,
            new_curr.num_of_matches,
            new_curr.id
          )

        cal_copies(new_copies, winning_cards, idx + 1)
      end
    end
  end

  def main() do
    board =
      File.read!("input.txt")
      |> String.split("\n")

    cards =
      board
      |> Enum.map(fn row ->
        splitAtId = String.split(row, ":")

        id = parse_id(Enum.at(splitAtId, 0))

        card =
          parse_row(Enum.at(splitAtId, 1))

        %{id: id, card: card}
      end)
      |> Enum.map(fn card ->
        {points, num_of_matches} = calculate_winning(card.card)
        %{id: card.id, card: card.card, w: points, num_of_matches: num_of_matches}
      end)
      |> Enum.filter(fn card ->
        card.id != nil
      end)

    total_points =
      cards
      |> Enum.reduce(0, fn card, acc ->
        case card do
          %{w: num} -> acc + num
          _ -> acc
        end
      end)

    %{
      total_points: total_points,
      total_cards:
        cal_copies(%{}, cards, 0)
        |> Enum.reduce(0, fn {_, num}, acc -> acc + num end)
    }
  end
end

IO.inspect(Day4.main())

File: go/day4/day4.go
package day4

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type card struct {
	cardId       int
	nums         map[int]int
	winning      map[int]int
	winningTotal int
	winningCards int
}

type cards []card
type cardCopies struct {
	copies map[int]int // cardId, num of copies
}

func (c *card) parseWinning(copies *cardCopies) {
	winningNums := 0
	for _, num := range c.winning {
		if _, ok := c.nums[num]; ok {
			if winningNums == 0 {
				c.winningTotal += 1
			} else {
				c.winningTotal *= 2
			}
			c.winningCards += 1

			winningNums += 1
		}
	}

	var copiesOfMe int
	if _, ok := copies.copies[c.cardId]; ok {
		copies.copies[c.cardId] += 1
		copiesOfMe = copies.copies[c.cardId]
	} else {
		copies.copies[c.cardId] = 1
		copiesOfMe = 1
	}

	if c.winningCards != 0 {
		for id := 1; id <= c.winningCards; id++ {
			idToAdd := id + c.cardId
			if _, ok := copies.copies[idToAdd]; ok {
				copies.copies[idToAdd] += copiesOfMe
			} else {
				copies.copies[idToAdd] = copiesOfMe
			}
		}
	}
}

func stringArrToNumMap(arr []string) map[int]int {
	nums := make(map[int]int)
	for _, num := range arr {
		if num != "" {
			n, err := strconv.Atoi(strings.TrimSpace(num))
			if err != nil {
				panic(err)
			}
			nums[n] = n
		}
	}

	return nums
}

func parseToCards(line string) card {
	c := card{}
	c.nums = make(map[int]int)
	c.winning = make(map[int]int)
	splitById := strings.Split(line, ":")
	cardId, err := strconv.Atoi(strings.TrimSpace(strings.Split(splitById[0], "Card")[1]))
	if err != nil {
		panic(err)
	}

	c.cardId = cardId

	splitNumsWinning := strings.Split(splitById[1], "|")

	c.nums = stringArrToNumMap(strings.Split(splitNumsWinning[0], " "))
	c.winning = stringArrToNumMap(strings.Split(splitNumsWinning[1], " "))

	return c
}

func day4() {
	fileName := "input.txt"
	file, err := os.Open(fileName)
	if err != nil {
		panic(err)
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	cards := make(cards, len(lines))
	cardCopies := cardCopies{copies: make(map[int]int)}
	totalPoints := 0
	totalCards := 0
	for i, line := range lines {
		cards[i] = parseToCards(line)
		cards[i].parseWinning(&cardCopies)
		totalPoints += cards[i].winningTotal
	}

	for _, copies := range cardCopies.copies {
		totalCards += copies
	}

	fmt.Printf("total points: %v\n", totalPoints)
	fmt.Printf("total cards: %v\n", totalCards)
}
